edward/inferences/ali.py

```diff
@@ -0,0 +1,86 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import six
+import tensorflow as tf
+
+from edward.inferences.gan_inference import GANInference
+from edward.util import get_session
+
+
+class ALI(GANInference):
+  """Adversarially Learned Inference (Dumoulin et al., 2016) or
+  Bidirectional Generative Adversarial Networks (Donahue et al., 2016)
+  for joint learning of generator and inference networks.
+
+  Works for the class of implicit (and differentiable) probabilistic
+  models. These models do not require a tractable density and assume
+  only a program that generates samples.
+  """
+  def __init__(self, *args, **kwargs):
+    """
+    Notes
+    -----
+    ``ALI`` matches a mapping from data to latent variables and a
+    mapping from latent variables to data through a joint
+    discriminator. The encoder approximates the posterior p(z|x)
+    when the network is stochastic.
+
+    In building the computation graph for inference, the
+    discriminator's parameters can be accessed with the variable scope
+    "Disc".
+    In building the computation graph for inference, the
+    encoder and decoder parameters can be accessed with the variable scope
+    "Gen".
+
+    Examples
+    --------
+    >>> with tf.variable_scope("Gen"):
+    >>>   xf = gen_data(z_ph)
+    >>>   zf = gen_latent(x_ph)
+    >>> inference = ed.ALI({xf: x_data, zf: z_samples}, discriminator)
```

Comment:
To match the interface of other inference methods, what about
```python
ed.ALI(latent_vars, data, discriminative_network)  # signature
ed.ALI({zf: qz}, {xf: x_data}, discriminator)  # example
```
Here the model's noise `zf` is binded to the inference network output `qz`, and the model's generated output `xf` is binded to data `x_data`.